### PR TITLE
WIP on Issue 94-add-remove-update-lines-in-the-databases

### PR DESCRIPTION
**Pull request by BhanuKandula5030**
What was changed?

- Added endpoint: `POST /page/{:id}/appendLine`
  - Description: Appends a line to the end of any existing lines in the specified Annotation Page identified by `{:id}`.
- Added endpoint: `POST /page/{:id}/prependLine`
  - Description: Adds a line to the top of the list in the specified Annotation Page identified by `{:id}`.
- Added endpoint: `POST /line/{:id}/after`
  - Description: Inserts a line in the list after the Line identified by `{:id}` (which is the hexString for a sibling Annotation).
- Added endpoint: `POST /line/{:id}/before`
  - Description: Inserts a line in the list before the Line identified by `{:id}`.
- Updated endpoint: `DELETE /line/{:id}/`
  - Description: Removes the specified Line identified by `{:id}` from the Annotation Page.
- Updated endpoint: `PUT /line/{:id}/`
  - Description: Replaces the content of the Line identified by `{:id}` with the payload of the request.

### CODE REVIEW

**Comment by BhanuKandula5030** [comment]
These are the Postman Screenshots of the response for the page requests append and prepend line 
pagepm1
pagepm2

**Review by thehabes — CHANGES_REQUESTED** [review]
Tests within your scope are failing.  Please make sure tests related to your scope (/line) are passing before you ask for review.  I do not see tests for the new endpoints you registered either.  All endpoints that intend to be merged with `development` must have some kind of useful test(s).  
> Your PR has been switched back to 'Draft' because you have not written these tests.

Found by running `npm run allTests` before looking at code.

image


image

File: line/index.mjs [hunk 1]
```diff
@@ -48,21 +49,78 @@ router.route('/:id')
...
+
+// I am using the below route for testing  and retriving the Annotation as mentioned as issue to check
+router.get('/:id/retrive', async (req, res) => {
+  const id = req.params.id
+  const siblingAnnotation = await logic.findingSiblingAnnotation(id)
+  return res.status(201).json(siblingAnnotation)
+})
+
+
+router.post('/:id/after', async (req, res) => {
```
Comment (thehabes) [inline line/index.mjs:61]: Add javadoc style documentation

```
/**
  * Please
*/
```

File: line/index.mjs [hunk 2]
```diff
@@ -48,21 +49,78 @@ router.route('/:id')
...
+      return res.status(404).json({ error: 'Line ${id} does not exist.' })
+    }
+    const response = await logic.insertLineinAnnotationPage(req.body, siblingAnnotation[0])
+    return res.status(201).json(response)
+  } catch (error) {
+    return res.status(500).json({ error: 'Internal server error' })
+  }
+})
+
+router.post('/:id/before', async (req, res) => {
```
Comment (thehabes) [inline line/index.mjs:75]: Add javadoc style documentation

```
/**
  * Please
*/
```

File: line/index.mjs [hunk 3]
```diff
@@ -48,21 +49,78 @@ router.route('/:id')
...
+      return res.status(404).json({ error: 'Line ${id} does not exist. '})
+    }
+    const response = await logic.insertLineBeforeAnnotation(req.body, siblingAnnotation[0])
+    return res.status(201).json(response)
+  } catch (error) {
+    return res.status(500).json({ error: 'Internal server error' })
+  }
+})
+ 
+router.delete('/:id', async (req, res) => {
```
Comment (thehabes) [inline line/index.mjs:89]: Add javadoc style documentation

```
/**
  * Please
*/
```

File: line/index.mjs [hunk 4]
```diff
@@ -48,21 +49,78 @@ router.route('/:id')
...
+      return res.status(404).json({ error: 'Line ${id} does not exist. '})
+    }
+    logic.deleteLineFromAnnotation(req.body, siblingAnnotation[0])
+    return res.status(204).json({"message": "Line deleted from annotion"})
+  } catch (error) {
+    return res.status(500).json({ error: 'Internal server error' })
+  }
+})
+
+router.put('/:id', async (req, res) => {
```
Comment (thehabes) [inline line/index.mjs:103]: Add javadoc style documentation

```
/**
  * Please
*/
```

**Review by thehabes — COMMENTED** [review]
I am running into a failed test when I use `npm run allTests`.  Look into this.
image

Make sure you get this to a functional state.  Here is the test I performed.

I generated Annotation Collection https://devstore.rerum.io/v1/id/6627fc13a1162cbdb9d34a5d

I generated Annotation Page https://devstore.rerum.io/v1/id/6627fc3d5cc3848690279f7c and put it into the collection.  This page has no `items[]` yet.

I performed http://localhost:3009/page/627fc3d5cc3848690279f7c/appendLine which said successful
image

However, no update was performed against Annotation Page https://devstore.rerum.io/v1/id/6627fc3d5cc3848690279f7c.  It does not have a `next` item in history, and also does not have any `items[]` even though I just added one.

It says it created two Annotations when I only provided one.  

image

In the first one, the @id http://example.org/anno1 is not resolvable.  This should be the devstore.rerum.io URI of the created annotation.

In the second one, there is no id which proves to me this Annotation was created.

**Comment by thehabes** [comment]
I had gotten the scope a little wrong.  After some conversation with @BhanuKandula5030 here's what we came up with for the scenario I was testing here.

```
{
  "@context": "http://www.w3.org/ns/anno.jsonld",
  "@id": "https://devstore.rerum.io/v1/id/6627fc13a1162cbdb9d34a5d"
  "type": "AnnotationCollection",
  "label": "Project - Layer Title",
  "creator": "thehabes",
  "total": 1,
  "first": "https://devstore.rerum.io/v1/id/6627fc3d5cc3848690279f7c",
  "last": "https://devstore.rerum.io/v1/id/6627fc3d5cc3848690279f7c",
  "partOf": "https://static.t-pen.org/6621c63a8168b7592a0bc627/project.json"
}

{
    "@context": "http://www.w3.org/ns/anno.jsonld",
    "@id":"https://devstore.rerum.io/v1/id/6627fc3d5cc3848690279f7c"
    "type": "AnnotationPage",
    "target" : "https://manifest/canvas/1",
    "next": "https://store.rerum.io/v1/id/page2",
    "items": [],
    "partOf": "https://devstore.rerum.io/v1/id/6627fc13a1162cbdb9d34a5d",
    "creator": "thehabes"
}

{
  "@context": "http://www.w3.org/ns/anno.jsonld",
  "@id": "https://devstore.rerum.io/v1/id/662801cbe8afc731e3b58e50",
  "creator": "thehabes",
  "type": "Annotation",
  "motivation": "supplementing",
  "body": {
    "type": "TextualBody",
    "value": "hello world",
    "format": "text/plain",
    "language": "la"
  },
  "target": "https://manifest/canvas/4#xywh=10,20,145,55"
}
```

What you have is an existing Annotation Page and an existing Annotation and they need to be put together.  Be aware of [how a PUT update works](https://store.rerum.io/v1/API.html#update) as you are not "overwriting" the object in place, but rather generating `history`.  This is important because it means you will have to do upstream changes to the Annotation Collection as well, but that will come next.  First, make sure /appendLine works with the provided Page and Annotation.

File: line/index.mjs (hunk 1 above)
Comment (BhanuKandula5030) [reply line/index.mjs:61]: Updated in the latest code .Thanks

File: line/index.mjs (hunk 3 above)
Comment (BhanuKandula5030) [reply line/index.mjs:89]: Updated in the latest code .Thanks

File: line/index.mjs (hunk 4 above)
Comment (BhanuKandula5030) [reply line/index.mjs:103]: Updated in the latest code .Thanks

**Comment by BhanuKandula5030** [comment]
Hi @thehabes 

I have made the failing test for the check in the previous code .Now I made the syntax corrections and made it pass .I am Attaching the test case check .All the tests are passing.

image

**Comment by BhanuKandula5030** [comment]
I have updated the append and prepend lone logics and used the https://devstore.rerum.io/v1/id/6627fc3d5cc3848690279f7c real annotation page provided and performed the tests. attaching the get requests of  before POST append line request  and after POST append line request screenshots where you see before as empty items and it got updated after append request  which in corresponds updated in the next in get requests.

before append
after append

**Comment by BhanuKandula5030** [comment]
I have updated prepend line logics and used the https://devstore.rerum.io/v1/id/6627fc3d5cc3848690279f7c real annotation page provided and performed the tests. attaching the get requests of before POST prepend line request and after POST prepend line request screenshots 
before prepend
after prepend

**Pull request closed without merging**